### DISCUSSION TITLE
How to use Discord's reconnection mechanism?

### DISCUSSION THREAD

**Discussion by linliyuan** [bwmarrin/discordgo#1501] [Q&A]
I found that there is relevant information in the documentation https://discord.com/developers/docs/topics/gateway#resuming, but there is no corresponding implementation method in the third-party package. Can this recovery mechanism be added? Currently, sessionID, sequence, and gateway are all private, which prevents me from obtaining the last sequence of websocket events. The adjustments needed are quite large. I saw that there were several merge requests related to this, what was the reason for finally closing them?
merge requests:
https://github.com/bwmarrin/discordgo/pull/1327
https://github.com/bwmarrin/discordgo/pull/1115

And the following piece of code is quite strange, the reconnection judgment is before ready, which inevitably leads to the sessionID and sequence in the newly created session being 0. Send Op 6 Resume Packet can only possibly be executed in the case of a reconnect.

![image](https://github.com/bwmarrin/discordgo/assets/37769385/60a487e4-25ed-4b69-9f16-96b81afe117b)

The current background is: the program cannot always run normally. If the service encounters an exception, I hope that the unprocessed data during the restart can be re-obtained through the recovery mechanism.

**Comment by jackulau** [comment]
are the id and sequence preserved?
